File: app/src/main/java/com/hetica/AutismoCordoba/TimerDescanso.java
package com.hetica.AutismoCordoba;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.os.Handler;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import java.util.Locale;

/**
 * The type Timer descanso.
 */
public class TimerDescanso extends AppCompatActivity {
    /**
     * The Time string.
     */
    String timeString;
    /**
     * The Actual.
     */
    String actual;
    /**
     * The Cuantas.
     */
    String cuantas;
    /**
     * The Bundle.
     */
    Bundle bundle;
    /**
     * The Asig.
     */
    String asig;

    private Ringtone r;

    private TextView mTextViewCountDown;
    private Button mButtonSiguiente;
    private Button mButtonFin;
    private Button Main;
    private long mStartTime;

    private int flag=1;

    private long mTimeLeftInMillis;
    private boolean mTimerRunning;
    private CountDownTimer mCountDownTimer;

    Intent siguiente;
    private long then;
    private int longClickDuration = 3000;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_timer_descanso);

        bundle = getIntent().getExtras();

        actual = bundle.getString("actAsig");
        cuantas = bundle.getString("numAsig");


        mStartTime=Long.parseLong("15")*60000;

        mTimeLeftInMillis=mStartTime;

        mTextViewCountDown = findViewById(R.id.text_view_count);

        Main=(Button)findViewById(R.id.button14);
        mButtonSiguiente = (Button)findViewById(R.id.button38);
        mButtonFin = (Button)findViewById(R.id.buttonFinDesc);
        mButtonSiguiente.setVisibility(View.INVISIBLE);
        mTextViewCountDown.setVisibility(View.VISIBLE);

        Main.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                if (event.getAction() == MotionEvent.ACTION_DOWN) {
                    then = (long) System.currentTimeMillis();
                } else if (event.getAction() == MotionEvent.ACTION_UP) {
                    if ((System.currentTimeMillis() - then) > longClickDuration) {
                        siguiente = new Intent(getBaseContext(), MainActivity.class);
                        mCountDownTimer.cancel();
                        mTimerRunning = false;
                        startActivity(siguiente);
                        System.out.println("Long Click has happened!");
                        return false;
                    } else {
                        /* Implement short click behavior here or do nothing */
                        System.out.println("Short Click has happened...");
                        return false;
                    }
                }
                return true;
            }
        });

        startTimer();
    }

    @Override
    public void onBackPressed() {

    }


    /**
     * Funcion que comienza y finaliza el temporizador del descanso
     *
     */
    private void startTimer() {
        mCountDownTimer = new CountDownTimer(mTimeLeftInMillis, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                mTimeLeftInMillis = millisUntilFinished;
                updateCountDownText();
            }

            @Override
            public void onFinish() {
                mTimerRunning = false;
                mButtonSiguiente.setVisibility(View.VISIBLE);
                mButtonFin.setVisibility(View.INVISIBLE);
                mTextViewCountDown.setText("00:00");
                //final MediaPlayer alarmSound = MediaPlayer.create(this, R.raw.algo);
                if (flag == 1) {
                    showNotification();
                }
                else{
                    pasar2();
                }
                //alarmSound.start();


            }
        }.start();

        mTimerRunning = true;

    }

    /**
     * Función que controla el temporizador
     *
     */
    private void updateCountDownText() {
        int minutes = (int) (mTimeLeftInMillis / 1000) / 60;
        int seconds = (int) (mTimeLeftInMillis / 1000) % 60;

        String timeLeftFormatted = String.format(Locale.getDefault(), "%02d:%02d", minutes, seconds);

        mTextViewCountDown.setText(timeLeftFormatted);



    }

    /**
     * Función que pasa a la siguiente actividad depende de la situación
     *
     * @param view the view
     */
    public void pasar2(){
        Bundle bundle2 = new Bundle();
        if ((cuantas.equalsIgnoreCase("2")) && actual.equalsIgnoreCase("1")) {
            Intent siguiente1 = new Intent(this, AsignaturaSiguiente.class);
            bundle2.putString("actAsig", "2");
            bundle2.putString("numAsig", cuantas);

            siguiente1.putExtras(bundle2);
            if(flag!=0) {
                r.stop();
            }
            startActivity(siguiente1);
        }
        if ((cuantas.equalsIgnoreCase("3") || cuantas.equalsIgnoreCase("4") || cuantas.equalsIgnoreCase("5")) && actual.equalsIgnoreCase("1")) {
            Intent siguiente1 = new Intent(this, AsignaturaDificil.class);
            bundle2.putString("actAsig", "2");
            bundle2.putString("numAsig", cuantas);

            siguiente1.putExtras(bundle2);
            if(flag!=0) {
                r.stop();
            }
            startActivity(siguiente1);
        }
        if ((cuantas.equalsIgnoreCase("3")) && actual.equalsIgnoreCase("2")) {
            Intent siguiente1 = new Intent(this, AsignaturaSiguiente.class);
            bundle2.putString("actAsig", "3");
            bundle2.putString("numAsig", cuantas);
            if(flag!=0) {
                r.stop();
            }
            siguiente1.putExtras(bundle2);
            startActivity(siguiente1);
        }
        if ((cuantas.equalsIgnoreCase("4") || cuantas.equalsIgnoreCase("5")) && actual.equalsIgnoreCase("2")) {
            Intent siguiente1 = new Intent(this, AsignaturaGustar.class);
            bundle2.putString("actAsig", "3");
            bundle2.putString("numAsig", cuantas);
            if(flag!=0) {
                r.stop();
            }
            siguiente1.putExtras(bundle2);
            startActivity(siguiente1);
        }
        if ((cuantas.equalsIgnoreCase("4") || cuantas.equalsIgnoreCase("5")) && actual.equalsIgnoreCase("3")) {
            Intent siguiente1 = new Intent(this, AsignaturaSiguiente.class);
            bundle2.putString("actAsig", "4");
            bundle2.putString("numAsig", cuantas);
            if(flag!=0) {
                r.stop();
            }
            siguiente1.putExtras(bundle2);
            startActivity(siguiente1);
        }
        if (cuantas.equalsIgnoreCase("5") && actual.equalsIgnoreCase("4")) {
            Intent siguiente1 = new Intent(this, AsignaturaSiguiente.class);
            bundle2.putString("actAsig", "5");
            bundle2.putString("numAsig", cuantas);
            if(flag!=0) {
                r.stop();
            }
            siguiente1.putExtras(bundle2);
            startActivity(siguiente1);
        }
        if (cuantas.equalsIgnoreCase(actual)){
            Intent siguiente1 = new Intent(this, MainActivity.class);
            if(flag!=0) {
                r.stop();
            }
            startActivity(siguiente1);
        }
    }

    public void pasar(View view){
        Bundle bundle2 = new Bundle();
        if ((cuantas.equalsIgnoreCase("2")) && actual.equalsIgnoreCase("1")) {
            Intent siguiente1 = new Intent(this, AsignaturaSiguiente.class);
            bundle2.putString("actAsig", "2");
            bundle2.putString("numAsig", cuantas);

            siguiente1.putExtras(bundle2);
            r.stop();
            startActivity(siguiente1);
        }
        if ((cuantas.equalsIgnoreCase("3") || cuantas.equalsIgnoreCase("4") || cuantas.equalsIgnoreCase("5")) && actual.equalsIgnoreCase("1")) {
            Intent siguiente1 = new Intent(this, AsignaturaDificil.class);
            bundle2.putString("actAsig", "2");
            bundle2.putString("numAsig", cuantas);

            siguiente1.putExtras(bundle2);
            r.stop();
            startActivity(siguiente1);
        }
        if ((cuantas.equalsIgnoreCase("3")) && actual.equalsIgnoreCase("2")) {
            Intent siguiente1 = new Intent(this, AsignaturaSiguiente.class);
            bundle2.putString("actAsig", "3");
            bundle2.putString("numAsig", cuantas);
            r.stop();
            siguiente1.putExtras(bundle2);
            startActivity(siguiente1);
        }
        if ((cuantas.equalsIgnoreCase("4") || cuantas.equalsIgnoreCase("5")) && actual.equalsIgnoreCase("2")) {
            Intent siguiente1 = new Intent(this, AsignaturaGustar.class);
            bundle2.putString("actAsig", "3");
            bundle2.putString("numAsig", cuantas);
            r.stop();
            siguiente1.putExtras(bundle2);
            startActivity(siguiente1);
        }
        if ((cuantas.equalsIgnoreCase("4") || cuantas.equalsIgnoreCase("5")) && actual.equalsIgnoreCase("3")) {
            Intent siguiente1 = new Intent(this, AsignaturaSiguiente.class);
            bundle2.putString("actAsig", "4");
            bundle2.putString("numAsig", cuantas);
            r.stop();
            siguiente1.putExtras(bundle2);
            startActivity(siguiente1);
        }
        if (cuantas.equalsIgnoreCase("5") && actual.equalsIgnoreCase("4")) {
            Intent siguiente1 = new Intent(this, AsignaturaSiguiente.class);
            bundle2.putString("actAsig", "5");
            bundle2.putString("numAsig", cuantas);
            r.stop();
            siguiente1.putExtras(bundle2);
            startActivity(siguiente1);
        }
        if (cuantas.equalsIgnoreCase(actual)){
            Intent siguiente1 = new Intent(this, MainActivity.class);
            r.stop();
            startActivity(siguiente1);
        }
    }

    public void showNotification(){

        try {
            Uri notification = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM);
            r = RingtoneManager.getRingtone(getApplicationContext(), notification);
            r.play();
            Handler handler = new Handler();
            handler.postDelayed(new Runnable() {
                public void run() {
                    r.stop();
                }
            }, 60000);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void Finalizar(View view){

        mCountDownTimer.cancel();
        mTimerRunning = false;
        flag = 0;
        mTimeLeftInMillis=1000;
        startTimer();
        mButtonFin.setVisibility(View.INVISIBLE);

    }
}

File: app/src/main/java/com/hetica/AutismoCordoba/TimerDescanso.kt
package com.hetica.AutismoCordoba

import android.annotation.SuppressLint
import android.content.Intent
import android.media.Ringtone
import android.media.RingtoneManager
import android.os.Bundle
import android.os.CountDownTimer
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.util.Log
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import java.util.Locale

/**
 * The type Timer descanso.
 */
class TimerDescanso : AppCompatActivity() {
    /**
     * The Time string.
     */
    var timeString: String? = null

    /**
     * The Actual.
     */
    var actual: String? = null

    /**
     * The Cuantas.
     */
    var cuantas: String? = null

    /**
     * The Bundle.
     */
    var bundle: Bundle? = null

    /**
     * The Asig.
     */
    var asig: String? = null
    private var r: Ringtone? = null
    private var mTextViewCountDown: TextView? = null
    private var mButtonSiguiente: Button? = null
    private var mButtonFin: Button? = null
    private var Main: Button? = null
    private var mStartTime: Long = 0
    private var flag = 1
    private var mTimeLeftInMillis: Long = 0
    private var mTimerRunning = false
    private var mCountDownTimer: CountDownTimer? = null
    var siguiente: Intent? = null
    private var then: Long = 0
    private val longClickDuration = 3000
    private val handler = Handler(Looper.getMainLooper())
    private val delayMillis = 3000L // 3 segundos
    private var isLongPressFired = false
    @SuppressLint("ClickableViewAccessibility")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_timer_descanso)
        bundle = intent.extras
        actual = bundle!!.getString("actAsig")
        cuantas = bundle!!.getString("numAsig")
        mStartTime = "15".toLong() * 60000
        mTimeLeftInMillis = mStartTime
        mTextViewCountDown = findViewById(R.id.text_view_count)
        Main = findViewById<View>(R.id.button14) as Button
        //mButtonSiguiente = findViewById<View>(R.id.button38) as Button
        mButtonFin = findViewById<View>(R.id.buttonFinDesc) as Button
        //mButtonSiguiente!!.visibility = View.INVISIBLE
        mTextViewCountDown?.visibility = View.VISIBLE

        main()
        startTimer()
    }

    override fun onBackPressed() {}
    @SuppressLint("ClickableViewAccessibility")
    private fun main() {
        val gestureDetector = GestureDetector(this, object : GestureDetector.SimpleOnGestureListener() {
            override fun onLongPress(e: MotionEvent) {
                if (isLongPressFired) {
                    return
                }
                isLongPressFired = true

                // Usamos un Handler para retrasar la apertura de la actividad EditarCalendario
                handler.postDelayed({
                    val intent = Intent(this@TimerDescanso, MainActivity::class.java)
                    startActivity(intent)
                }, delayMillis)
            }
        })

        Main?.setOnTouchListener { _, event ->
            gestureDetector.onTouchEvent(event)
            if (event.action == MotionEvent.ACTION_UP || event.action == MotionEvent.ACTION_CANCEL) {
                isLongPressFired = false
                // Si se libera el botón antes del tiempo de espera, cancelamos el Handler
                handler.removeCallbacksAndMessages(null)
            }
            true
        }
    }
    /**
     * Funcion que comienza y finaliza el temporizador del descanso
     *
     */
    private fun startTimer() {
        mCountDownTimer = object : CountDownTimer(mTimeLeftInMillis, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                mTimeLeftInMillis = millisUntilFinished
                updateCountDownText()
            }

            override fun onFinish() {
                mTimerRunning = false
                //mButtonSiguiente!!.visibility = View.VISIBLE
                mButtonFin!!.visibility = View.INVISIBLE
                mTextViewCountDown?.text = "00:00"

                //final MediaPlayer alarmSound = MediaPlayer.create(this, R.raw.algo);
                if (flag == 1) {
                    showNotification()
                } else {
                    pasar2()
                }
                //alarmSound.start();
            }
        }.start()
        mTimerRunning = true
    }

    /**
     * Función que controla el temporizador
     *
     */
    private fun updateCountDownText() {
        val minutes = (mTimeLeftInMillis / 1000).toInt() / 60
        val seconds = (mTimeLeftInMillis / 1000).toInt() % 60
        val timeLeftFormatted = String.format(Locale.getDefault(), "%02d:%02d", minutes, seconds)
        mTextViewCountDown!!.text = timeLeftFormatted
    }

    /**
     * Función que pasa a la siguiente actividad depende de la situación
     *
     * @param view the view
     */
    fun pasar2() {
        val bundle2 = Bundle()
        if (cuantas.equals("2", ignoreCase = true) && actual.equals("1", ignoreCase = true)) {
            val siguiente1 = Intent(this, AsignaturaSiguiente::class.java)
            bundle2.putString("actAsig", "2")
            bundle2.putString("numAsig", cuantas)
            siguiente1.putExtras(bundle2)
            if (flag != 0) {
                r!!.stop()
            }
            startActivity(siguiente1)
        }
        if ((cuantas.equals("3", ignoreCase = true) || cuantas.equals("4", ignoreCase = true) || cuantas.equals("5", ignoreCase = true)) && actual.equals("1", ignoreCase = true)) {
            val siguiente1 = Intent(this, AsignaturaDificil::class.java)
            bundle2.putString("actAsig", "2")
            bundle2.putString("numAsig", cuantas)
            siguiente1.putExtras(bundle2)
            if (flag != 0) {
                r!!.stop()
            }
            startActivity(siguiente1)
        }
        if (cuantas.equals("3", ignoreCase = true) && actual.equals("2", ignoreCase = true)) {
            val siguiente1 = Intent(this, AsignaturaSiguiente::class.java)
            bundle2.putString("actAsig", "3")
            bundle2.putString("numAsig", cuantas)
            if (flag != 0) {
                r!!.stop()
            }
            siguiente1.putExtras(bundle2)
            startActivity(siguiente1)
        }
        if ((cuantas.equals("4", ignoreCase = true) || cuantas.equals("5", ignoreCase = true)) && actual.equals("2", ignoreCase = true)) {
            val siguiente1 = Intent(this, AsignaturaGustar::class.java)
            bundle2.putString("actAsig", "3")
            bundle2.putString("numAsig", cuantas)
            if (flag != 0) {
                r!!.stop()
            }
            siguiente1.putExtras(bundle2)
            startActivity(siguiente1)
        }
        if ((cuantas.equals("4", ignoreCase = true) || cuantas.equals("5", ignoreCase = true)) && actual.equals("3", ignoreCase = true)) {
            val siguiente1 = Intent(this, AsignaturaSiguiente::class.java)
            bundle2.putString("actAsig", "4")
            bundle2.putString("numAsig", cuantas)
            if (flag != 0) {
                r!!.stop()
            }
            siguiente1.putExtras(bundle2)
            startActivity(siguiente1)
        }
        if (cuantas.equals("5", ignoreCase = true) && actual.equals("4", ignoreCase = true)) {
            val siguiente1 = Intent(this, AsignaturaSiguiente::class.java)
            bundle2.putString("actAsig", "5")
            bundle2.putString("numAsig", cuantas)
            if (flag != 0) {
                r!!.stop()
            }
            siguiente1.putExtras(bundle2)
            startActivity(siguiente1)
        }
        if (cuantas.equals(actual, ignoreCase = true)) {
            val siguiente1 = Intent(this, MainActivity::class.java)
            if (flag != 0) {
                r!!.stop()
            }
            startActivity(siguiente1)
        }
    }

    fun pasar(view: View?) {
        val bundle2 = Bundle()
        if (cuantas.equals("2", ignoreCase = true) && actual.equals("1", ignoreCase = true)) {
            val siguiente1 = Intent(view!!.context, AsignaturaSiguiente::class.java)
            bundle2.putString("actAsig", "2")
            bundle2.putString("numAsig", cuantas)
            siguiente1.putExtras(bundle2)
            r!!.stop()
            startActivity(siguiente1)
        }
        if ((cuantas.equals("3", ignoreCase = true) || cuantas.equals("4", ignoreCase = true) || cuantas.equals("5", ignoreCase = true)) && actual.equals("1", ignoreCase = true)) {
            val siguiente1 = Intent(this, AsignaturaDificil::class.java)
            bundle2.putString("actAsig", "2")
            bundle2.putString("numAsig", cuantas)
            siguiente1.putExtras(bundle2)
            r!!.stop()
            startActivity(siguiente1)
        }
        if (cuantas.equals("3", ignoreCase = true) && actual.equals("2", ignoreCase = true)) {
            val siguiente1 = Intent(this, AsignaturaSiguiente::class.java)
            bundle2.putString("actAsig", "3")
            bundle2.putString("numAsig", cuantas)
            r!!.stop()
            siguiente1.putExtras(bundle2)
            startActivity(siguiente1)
        }
        if ((cuantas.equals("4", ignoreCase = true) || cuantas.equals("5", ignoreCase = true)) && actual.equals("2", ignoreCase = true)) {
            val siguiente1 = Intent(this, AsignaturaGustar::class.java)
            bundle2.putString("actAsig", "3")
            bundle2.putString("numAsig", cuantas)
            r!!.stop()
            siguiente1.putExtras(bundle2)
            startActivity(siguiente1)
        }
        if ((cuantas.equals("4", ignoreCase = true) || cuantas.equals("5", ignoreCase = true)) && actual.equals("3", ignoreCase = true)) {
            val siguiente1 = Intent(this, AsignaturaSiguiente::class.java)
            bundle2.putString("actAsig", "4")
            bundle2.putString("numAsig", cuantas)
            r!!.stop()
            siguiente1.putExtras(bundle2)
            startActivity(siguiente1)
        }
        if (cuantas.equals("5", ignoreCase = true) && actual.equals("4", ignoreCase = true)) {
            val siguiente1 = Intent(this, AsignaturaSiguiente::class.java)
            bundle2.putString("actAsig", "5")
            bundle2.putString("numAsig", cuantas)
            r!!.stop()
            siguiente1.putExtras(bundle2)
            startActivity(siguiente1)
        }
        if (cuantas.equals(actual, ignoreCase = true)) {
            val siguiente1 = Intent(this, MainActivity::class.java)
            r!!.stop()
            startActivity(siguiente1)
        }
    }

    fun showNotification() {
        try {
            val notification = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM)
            r = RingtoneManager.getRingtone(applicationContext, notification)
            r?.play()

            val handlerThread = HandlerThread("StopRingtoneThread")
            handlerThread.start()

            val handler = Handler(handlerThread.looper)
            handler.postDelayed({
                r?.stop()
                handlerThread.quitSafely()
            }, 60000)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }


    fun Finalizar(view: View?) {
        Log.d("Finalizando", "Finalizando usando $view")
        mCountDownTimer!!.cancel()
        mTimerRunning = false
        flag = 0
        mTimeLeftInMillis = 1000
        startTimer()
        mButtonFin!!.visibility = View.INVISIBLE
    }
}